File: Codes from the AI-assistants/Old Experiments/[CLBG] k-nucleotide/Basic Prompt - Normal Mode/KNucleotide-copilot-basic-normal.cpp
import sys
import re
from collections import Counter

def read_fasta_input():
    data = sys.stdin.read()
    match = re.search(r'>THREE\s+([\w\n]+)', data)
    if not match:
        raise ValueError("DNA sequence THREE not found.")
    sequence = ''.join(match.group(1).split()).upper()
    return sequence

def count_kmers(seq, k):
    counter = Counter()
    for i in range(len(seq) - k + 1):
        kmer = seq[i:i+k]
        counter[kmer] += 1
    return counter

def print_frequencies(counter, total):
    sorted_items = sorted(
        counter.items(),
        key=lambda item: (-item[1], item[0])
    )
    for kmer, count in sorted_items:
        freq = (count / total) * 100
        print(f"{kmer} {freq:.3f}")

def print_specific_kmers(counter, kmers):
    for kmer in kmers:
        print(f"{counter.get(kmer, 0)}\t{kmer}")

def main():
    seq = read_fasta_input()

    # Frequencies for k = 1 and 2
    for k in [1, 2]:
        counter = count_kmers(seq, k)
        total = sum(counter.values())
        print_frequencies(counter, total)
        print()

    # Specific sequences
    for k, kmers in [
        (3, ['GGT']),
        (4, ['GGTA']),
        (6, ['GGTATT']),
        (12, ['GGTATTTTAATT']),
        (18, ['GGTATTTTAATTTATAGT']),
    ]:
        counter = count_kmers(seq, k)
        print_specific_kmers(counter, kmers)

if __name__ == '__main__':
    main()

File: Codes from the AI-assistants/Old Experiments/[Experimental] k-nucleotide/Basic Prompt - Normal Mode/KNucleotide-copilot-basic-normal.py
from collections import Counter

def k_nucleotide_frequencies(dna_sequence: str, k: int) -> dict:
    if k <= 0 or k > len(dna_sequence):
        return {}

    # Extract all k-length substrings
    k_mers = (dna_sequence[i:i + k] for i in range(len(dna_sequence) - k + 1))
    
    # Count using native Counter
    frequencies = Counter(k_mers)
    
    # Optionally sort by frequency or alphabetically
    return dict(sorted(frequencies.items(), key=lambda x: (-x[1], x[0])))

dna = "AGCTAGCTAGCTA"
k = 3

result = k_nucleotide_frequencies(dna, k)
for kmer, count in result.items():
    print(f"{kmer}: {count}")
